collector/internal/lifecycle/manager.go

```diff
@@ -142,10 +145,37 @@ func (lm *manager) processEvents(ctx context.Context) error {
 				return err
 			}
 
+			lm.notifyFunctionInvoked()
+
 			err = lm.listener.Wait(ctx, res.RequestID)
 			if err != nil {
 				lm.logger.Error("problem waiting for platform.runtimeDone event", zap.Error(err), zap.String("requestID", res.RequestID))
 			}
+
+			// Check other components are ready before allowing the freezing of the environment.
```

Comment:
@tylerbenson  - here's a comment to explain that we notifier other components that the function has finished and therefore need to be ready to freeze.

The new code I've added hooks into the existing lifecycle code,

Reply:
I don't have a good understanding of the extension lifecycle.  If the lambda function returns and starts waiting for the next event, are you sure it waits for events to finish processing before freezing? Is this documented somewhere? How did you verify this behavior?

Reply:
Lifecycle docs say it waits until the extension also calls `Next`.

But does the lambda function get the next event only after all extensions have called `Next`? If that were the case then wouldn't it still be waiting until the extension finishes, unless the extension calls `Next` before doing the export, but then it might be frozen before actually exporting.

Reply:
@tylerbenson according to the lifecycle documentation (https://docs.aws.amazon.com/lambda/latest/dg/runtimes-extensions-api.html#runtimes-extensions-api-lifecycle)

> Each phase starts with an event from Lambda to the runtime and to all registered extensions. The runtime and each extension signal completion by sending a Next API request. Lambda freezes the execution environment when each process has completed and there are no pending events.

@tsloughter the lambda runtime will only get the next invocation once all extensions have also signalled they are ready. 
The sequence of events is:
1. Lambda and Extensions receive `Next` event
2. Lambda function completes and signals it's ready for the next event.
3. Collector receives notification from Telemetry API that the function has finished.
4. Decouple processor finishes forwarding data.
5. Collector signals it's ready for the next event.
6. Back to 1.
 
Also see the "Invoke phase" in the same link:

> After receiving the function response from the runtime, Lambda returns the response to the client, even if extensions are still running.

and 

> The Invoke phase ends after the runtime and all extensions signal that they are done by sending a Next API request.

As to the comment about verifying this behaviour - from Cloudwatch logs to confirm what's documented.

Reply:
Thanks for the detailed response.

Reply:
Ok, cool, this is great.